crates/astria-sequencer/src/asset/query.rs

```diff
@@ -0,0 +1,101 @@
+use anyhow::Context as _;
+use astria_core::{
+    primitive::v1::asset,
+    protocol::abci::AbciErrorCode,
+};
+use cnidarium::Storage;
+use prost::Message as _;
+use tendermint::abci::{
+    request,
+    response,
+};
+
+use crate::{
+    asset::state_ext::StateReadExt as _,
+    state_ext::StateReadExt,
+};
+
+// Retrieve the full asset denomination given the asset ID.
+//
+// Example:
+// `abci-cli query --path=asset/denom/<DENOM_ID>`
+pub(crate) async fn denom_request(
+    storage: Storage,
+    request: request::Query,
+    params: Vec<(String, String)>,
+) -> response::Query {
+    use astria_core::protocol::asset::v1alpha1::DenomResponse;
+
+    // use the latest snapshot, as this is a lookup of id->denom
+    let snapshot = storage.latest_snapshot();
+    let asset_id = match preprocess_request(&params).await {
+        Ok(asset_id) => asset_id,
+        Err(err_rsp) => return err_rsp,
+    };
+
+    let height = match snapshot.get_block_height().await {
+        Ok(height) => height,
+        Err(err) => {
+            return response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+                log: format!("failed getting block height: {err:?}"),
+                ..response::Query::default()
+            };
+        }
+    };
+
+    let denom = match snapshot.get_ibc_asset(asset_id).await {
+        Ok(denom) => denom,
+        Err(err) => {
+            return response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+                log: format!("failed to retrieve denomination: {err:?}"),
+                ..response::Query::default()
+            };
+        }
+    };
+
+    let payload = DenomResponse {
+        height: height as u64,
+        denom: denom.clone(),
+    }
+    .into_raw()
+    .encode_to_vec()
+    .into();
+
+    let height = u32::try_from(height).expect("height must fit into a u32");
```

Comment:
Probably not really an issue, but this `try_from` is more restrictive than `Height::try_from` where it only requires the value to be not more than `i64::MAX`.

```suggestion
    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
```